Print associated types on opaque `impl Trait` types

compiler-errors opened the pull request:
This PR generalizes #91021, printing associated types for all opaque `impl Trait` types instead of just special-casing for future. 

before: 
```
error[E0271]: type mismatch resolving `<impl Iterator as Iterator>::Item == u32`
```

after:
```
error[E0271]: type mismatch resolving `<impl Iterator<Item = usize> as Iterator>::Item == u32`
```

---

Questions:
1. I'm kinda lost in binders hell with this one. Is all of the `rebind`ing necessary?
2. Is there a map collection type that will give me a stable iteration order? Doesn't seem like TraitRef is Ord, so I can't just sort later..
3. I removed the logic that suppresses printing generator projection types. It creates outputs like this [gist](https://gist.github.com/compiler-errors/d6f12fb30079feb1ad1d5f1ab39a3a8d). Should I put that back?
4. I also added spaces between traits, `impl A+B` -> `impl A + B`. I quite like this change, but is there a good reason to keep it like that?

r? @estebank

estebank commented:
> I also added spaces between traits, impl A+B -> impl A + B. I quite like this change, but is there a good reason to keep it like that?

Can we move that to a separate PR? I can approve and merge that quickly.

> Is there a map collection type that will give me a stable iteration order? Doesn't seem like TraitRef is Ord, so I can't just sort later.

You can use `sort_by` to get around that issue. Any container that does stable sorted iteration will have that problem :-/

> I removed the logic that suppresses printing generator projection types. It creates outputs like this gist. Should I put that back?

Some of the output is getting quite verbose :(

The one that stands out to me in particular is for closures: multiple `Fn*` traits are implemented for the type, and now we show them all, instead of showing only the one that is relevant. The nested `Future` cases also look... unwieldy. It might be we do want to show them in some cases, but I'm slightly concerned. I need to think about what we can do here a bit more in depth.

estebank commented on compiler/rustc_middle/src/ty/print/pretty.rs:
```diff
@@ -647,68 +647,97 @@ pub trait PrettyPrinter<'tcx>:
+
+                                // Append to trait_ref's projection list if it exists, otherwise
+                                // insert a new one.
+                                if let Some((_, assoc_items)) =
+                                    traits.iter_mut().find(|(t, _)| t == &trait_ref)
                                 {
-                                    is_future = true;
-                                    continue;
+                                    assoc_items.push(proj_ty);
+                                } else {
+                                    traits.push((trait_ref, vec![proj_ty]));
                                 }
```
All of this can be replaced by using the `.entry()` API instead.

compiler-errors replied:
Yeah, that's what I was using before I hit issues with sorting.

compiler-errors replied:
To make myself a bit more clear, I was using an HashMap here before and when I iterated over the entries during the printing phase, the test outputs were randomized across runs. I tried to fix this by then collecting back into a Vec only after iteration, but I didn't know what to `sort_by` since all of my keys aren't `Ord`.

compiler-errors commented:
> You can use sort_by to get around that issue

Sort by what though? PolyTraitRef (or TraitRef) which I would want to use as the key to this mapping is not Ord. Should I sort by the ref's DefId and its substs?

compiler-errors commented:
> Can we move that to a separate PR? I can approve and merge that quickly.

Can do.

estebank commented on src/test/ui/lint/opaque-ty-ffi-unsafe.stderr:
```diff
@@ -1,4 +1,4 @@
-error: `extern` block uses type `impl Fn<()>`, which is not FFI-safe
+error: `extern` block uses type `impl Fn<()> + FnOnce<(), Output = ()>`, which is not FFI-safe
```
This is what I was mentioning. I think that we can look at the traits to see that one implies the other to show only one at a time, without special casing closures.

compiler-errors replied:
Yeah so this is printing because we have two predicates here `Fn<()>` and `<Self as FnOnce<()>>::Output == ()`. The first one makes us print `Fn<()>` and the second makes us print `FnOnce<(), Output = ()>`.

I actually would be interested in special casing just `Fn`/`FnMut`/`FnOnce` and fixing it up further so it prints `impl Fn() -> ()` instead.

estebank replied:
You can keep track of the traits for the predicates you've seen already and see if the new predicate's trait is part of the existing ones' super-traits. Special casing sounds reasonable-ish, but I would prefer it if we could just make it work for the general case :)

Regardless, dealing with the `<_ as async {}>::Return` types in a nicer way (even if it is showing them as `_`) has priority over this.

compiler-errors replied:
> You can keep track of the traits for the predicates you've seen already

Is there any ordering enforced for the list of predicates that we're iterating over here, though? 

And I'm pretty sure this only prints supertraits if you're bounding an associated type on that supertrait, so in practice we really only see this with `Fn*` family of traits I think. (Because `impl Fn() -> Foo` desugars to `impl Fn<()> + FnOnce<(), Output=Foo>`, or similar, right?)

estebank replied:
Closures is gonna be the most common thing people will see, but a crate can write their own API structure that does the same thing. If closures are special cased, any user of those (potential, as none come to mind) crates could get confused.

estebank commented on src/test/ui/impl-trait/issues/issue-78722.stderr:
```diff
@@ -15,7 +15,7 @@ LL | pub const fn from_generator<T>(gen: T) -> impl Future<Output = T::Return>
    |
    = note: expected opaque type `impl Future<Output = u8>`
-              found opaque type `impl Future`
+              found opaque type `impl Future<Output = <[static generator@$DIR/issue-78722.rs:12:26: 12:31] as Generator<ResumeTy>>::Return>`
```
I think though that we *should* special case `Generator` and call it something like `[async@$DIR/issue-78722.rs:12:20: 12:31]` so it looks like:

```
found opaque type `impl Future<Output = <[async@$DIR/issue-78722.rs:12:26: 12:31]>`
```

What do you think?

compiler-errors replied:
Hmm.. Yeah, I could see us special-casing it. I wouldn't mind printing `[async@...]` for the whole `impl Future` type..

... unless there's a way of normalizing that `<_ as Generator>::Return` type in the code here. That would be the best.

estebank replied:
If you make `Generator::Return` a `lang_item` and query it here to unconditionally replace it with `_` or `[async result]`, what is the diff in the tests? We might want to go that route instead, depending on how much real code is affected. I suspect that this type isn't being shown much today because it most commonly appears only as part of `Future::Output`, so it wouldn't be a regression to hide it.

bors commented:
:umbrella: The latest upstream changes (presumably #91104) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

compiler-errors commented:
I updated the code: 
* Replace those noisy generator projection ty with `[async output]`. I would be fine with also replacing it with `_`.
* Add logic to special-case print `impl Fn(A, B, C) -> D` for `Fn`/`FnMut`/`FnOnce` in the cases where the predicates work out.

And smaller things:
* Derive Ord for TraitRef so I could use `BTreeMap` instead of `Vec` (which stably prints error outputs now). 
* Added lang-item for `generator_return`.
* Coped and simplified [`supertraits`](https://doc.rust-lang.org/beta/nightly-rustc/src/rustc_infer/traits/util.rs.html#277-282) method from `rustc_infer` since we can't access that from `rustc_middle`, but we need it for `impl Fn` printing.

What do you think @estebank ?

estebank commented on compiler/rustc_middle/src/ty/print/pretty.rs:
```diff
@@ -826,6 +753,218 @@ pub trait PrettyPrinter<'tcx>:
+                        p!(print(ty));
+                    }
+
+                    p!(") -> ", print(return_ty), write("{}", if paren_needed { ")" } else { "" }));
```
Could we special case `impl Fn()` when `return_ty` is `()`?

compiler-errors replied:
Can do.

estebank commented on src/test/ui/impl-trait/issues/issue-86201.stderr:
```diff
@@ -7,7 +7,7 @@ LL |
    |                             ^^^^^^^ expected opaque type, found fn item
    |
-   = note: expected opaque type `impl Fn<()>`
+   = note: expected opaque type `impl Fn<()> + FnOnce<()>`
```
Noticed this, a bit odd because it seems you were detecting this case already

compiler-errors replied:
So the reason this one doesn't rewrite itself into `impl Fn() -> T` is because the opaque type has no return type constraint. I'll fix it to just go back to `impl Fn<()>` (very easy to do) or we could print `-> _` if you think that would look better!

estebank replied:
Either approach would be fine. Do whichever works best for you.

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-12`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/4303331237?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/4c6194ff5ece056db2e0cc4087ffdd382f633915/checks/4303331237/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
Some tests failed in compiletest suite=mir-opt mode=mir-opt host=x86_64-unknown-linux-gnu target=x86_64-unknown-linux-gnu
.................................i................................
failures:

---- [mir-opt] mir-opt/inline/issue-78442.rs stdout ----
5       debug _baz => _1;                    // in scope 0 at $DIR/issue-78442.rs:9:5: 9:9
6       let mut _0: ();                      // return place in scope 0 at $DIR/issue-78442.rs:10:3: 10:3
7       let _2: ();                          // in scope 0 at $DIR/issue-78442.rs:11:5: 11:17
- -     let mut _3: &impl Fn() -> ();        // in scope 0 at $DIR/issue-78442.rs:11:5: 11:15
- -     let _4: impl Fn() -> ();             // in scope 0 at $DIR/issue-78442.rs:11:5: 11:15
+ -     let mut _3: &impl Fn();              // in scope 0 at $DIR/issue-78442.rs:11:5: 11:15
+ -     let _4: impl Fn();                   // in scope 0 at $DIR/issue-78442.rs:11:5: 11:15
10 +     let mut _3: &fn() {foo};             // in scope 0 at $DIR/issue-78442.rs:11:5: 11:15
11 +     let _4: fn() {foo};                  // in scope 0 at $DIR/issue-78442.rs:11:5: 11:15
12       let mut _5: ();                      // in scope 0 at $DIR/issue-78442.rs:11:5: 11:17

18           _4 = hide_foo() -> [return: bb1, unwind: bb4]; // scope 0 at $DIR/issue-78442.rs:11:5: 11:15
19                                            // mir::Constant
20                                            // + span: $DIR/issue-78442.rs:11:5: 11:13
-                                            // + literal: Const { ty: fn() -> impl Fn() -> () {hide_foo}, val: Value(Scalar(<ZST>)) }
+                                            // + literal: Const { ty: fn() -> impl Fn() {hide_foo}, val: Value(Scalar(<ZST>)) }
23   
24       bb1: {


25           _3 = &_4;                        // scope 0 at $DIR/issue-78442.rs:11:5: 11:15
26           StorageLive(_5);                 // scope 0 at $DIR/issue-78442.rs:11:5: 11:17
27           nop;                             // scope 0 at $DIR/issue-78442.rs:11:5: 11:17
-           _2 = <impl Fn() -> () as Fn<()>>::call(move _3, move _5) -> [return: bb2, unwind: bb4]; // scope 0 at $DIR/issue-78442.rs:11:5: 11:17
+           _2 = <impl Fn() as Fn<()>>::call(move _3, move _5) -> [return: bb2, unwind: bb4]; // scope 0 at $DIR/issue-78442.rs:11:5: 11:17
29                                            // mir::Constant
30                                            // + span: $DIR/issue-78442.rs:11:5: 11:15
-                                            // + literal: Const { ty: for<'r> extern "rust-call" fn(&'r impl Fn() -> (), ()) -> <impl Fn() -> () as std::ops::FnOnce<()>>::Output {<impl Fn() -> () as std::ops::Fn<()>>::call}, val: Value(Scalar(<ZST>)) }
+                                            // + literal: Const { ty: for<'r> extern "rust-call" fn(&'r impl Fn(), ()) -> <impl Fn() as std::ops::FnOnce<()>>::Output {<impl Fn() as std::ops::Fn<()>>::call}, val: Value(Scalar(<ZST>)) }
33   
34       bb2: {


thread '[mir-opt] mir-opt/inline/issue-78442.rs' panicked at 'Actual MIR output differs from expected MIR output /checkout/src/test/mir-opt/inline/issue_78442.bar.RevealAll.diff', src/tools/compiletest/src/runtest.rs:3357:25


failures:
    [mir-opt] mir-opt/inline/issue-78442.rs
    [mir-opt] mir-opt/inline/issue-78442.rs

test result: FAILED. 162 passed; 1 failed; 3 ignored; 0 measured; 0 filtered out; finished in 5.23s



command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0-tools-bin/compiletest" "--compile-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib" "--run-lib-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/lib/rustlib/x86_64-unknown-linux-gnu/lib" "--rustc-path" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "--src-base" "/checkout/src/test/mir-opt" "--build-base" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/mir-opt" "--stage-id" "stage2-x86_64-unknown-linux-gnu" "--suite" "mir-opt" "--mode" "mir-opt" "--target" "x86_64-unknown-linux-gnu" "--host" "x86_64-unknown-linux-gnu" "--llvm-filecheck" "/usr/lib/llvm-12/bin/FileCheck" "--nodejs" "/usr/bin/node" "--host-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--target-rustcflags" "-Crpath -O -Cdebuginfo=0  -Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "--docck-python" "/usr/bin/python3" "--lldb-python" "/usr/bin/python3" "--gdb" "/usr/bin/gdb" "--quiet" "--llvm-version" "12.0.0" "--llvm-components" "aarch64 aarch64asmparser aarch64codegen aarch64desc aarch64disassembler aarch64info aarch64utils aggressiveinstcombine all all-targets amdgpu amdgpuasmparser amdgpucodegen amdgpudesc amdgpudisassembler amdgpuinfo amdgpuutils analysis arm armasmparser armcodegen armdesc armdisassembler arminfo armutils asmparser asmprinter avr avrasmparser avrcodegen avrdesc avrdisassembler avrinfo binaryformat bitreader bitstreamreader bitwriter bpf bpfasmparser bpfcodegen bpfdesc bpfdisassembler bpfinfo cfguard codegen core coroutines coverage debuginfocodeview debuginfodwarf debuginfogsym debuginfomsf debuginfopdb demangle dlltooldriver dwarflinker engine executionengine extensions filecheck frontendopenacc frontendopenmp fuzzmutate globalisel hellonew hexagon hexagonasmparser hexagoncodegen hexagondesc hexagondisassembler hexagoninfo instcombine instrumentation interfacestub interpreter ipo irreader jitlink lanai lanaiasmparser lanaicodegen lanaidesc lanaidisassembler lanaiinfo libdriver lineeditor linker lto mc mca mcdisassembler mcjit mcparser mips mipsasmparser mipscodegen mipsdesc mipsdisassembler mipsinfo mirparser msp430 msp430asmparser msp430codegen msp430desc msp430disassembler msp430info native nativecodegen nvptx nvptxcodegen nvptxdesc nvptxinfo objcarcopts object objectyaml option orcjit orcshared orctargetprocess passes perfjitevents powerpc powerpcasmparser powerpccodegen powerpcdesc powerpcdisassembler powerpcinfo profiledata remarks riscv riscvasmparser riscvcodegen riscvdesc riscvdisassembler riscvinfo runtimedyld scalaropts selectiondag sparc sparcasmparser sparccodegen sparcdesc sparcdisassembler sparcinfo support symbolize systemz systemzasmparser systemzcodegen systemzdesc systemzdisassembler systemzinfo tablegen target textapi transformutils vectorize webassembly webassemblyasmparser webassemblycodegen webassemblydesc webassemblydisassembler webassemblyinfo windowsmanifest x86 x86asmparser x86codegen x86desc x86disassembler x86info xcore xcorecodegen xcoredesc xcoredisassembler xcoreinfo xray" "--system-llvm" "--cc" "" "--cxx" "" "--cflags" "" "--adb-path" "adb" "--adb-test-dir" "/data/tmp/work" "--android-cross-path" "" "--channel" "nightly" "--color" "always"


Build completed unsuccessfully in 0:13:04
```

</details>

estebank commented on src/test/ui/entry-point/imported_main_const_fn_item_type_forbidden.stderr:
```diff
@@ -21,7 +21,7 @@ LL |     type MainFn = impl Fn();
    |                             ^^^ expected opaque type, found fn item
    |
-   = note: expected opaque type `impl Fn<()>`
+   = note: expected opaque type `impl Fn()`
                   found fn item `fn() {bar}`
```
(not for this PR): These cases should have an additional `note` explaining the difference between `fn` and `Fn`, and what you can do to deal with that.

estebank commented:
@bors r+

bors commented:
:pushpin: Commit 9cc117930b02b86795946e6b8748166c17fd4f16 has been approved by `estebank`

<!-- @bors r=estebank 9cc117930b02b86795946e6b8748166c17fd4f16 -->
<!-- homu: {"type":"Approved","sha":"9cc117930b02b86795946e6b8748166c17fd4f16","approver":"estebank"} -->

bors commented:
:hourglass: Testing commit 9cc117930b02b86795946e6b8748166c17fd4f16 with merge 26389d998e6ee0492f2c7f3eec661d37e9d6d841...
<!-- homu: {"type":"BuildStarted","head_sha":"9cc117930b02b86795946e6b8748166c17fd4f16","merge_sha":"26389d998e6ee0492f2c7f3eec661d37e9d6d841"} -->

rust-log-analyzer commented:
The job **`x86_64-gnu-tools`** failed! Check out the build log: [(web)](https://github.com/rust-lang-ci/rust/runs/4309413981?check_suite_focus=true) [(plain)](https://github.com/rust-lang-ci/rust/commit/26389d998e6ee0492f2c7f3eec661d37e9d6d841/checks/4309413981/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
......... (50/53)
...        (53/53)


/checkout/src/test/rustdoc-gui/toggle-click-deadspace.goml toggle-click-deadspace... FAILED
[ERROR] (line 8) Error: Evaluation failed: assert didn't fail: for command `assert-attribute-false: (".impl-items .rustdoc-toggle", {"open": ""})`



command did not execute successfully: "/node-v14.4.0-linux-x64/bin/node" "/checkout/src/tools/rustdoc-gui/tester.js" "--jobs" "16" "--doc-folder" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/rustdoc-gui/doc" "--tests-folder" "/checkout/src/test/rustdoc-gui"


Build completed unsuccessfully in 0:00:16
```

</details>

bors commented:
:broken_heart: Test failed - [checks-actions](https://github.com/rust-lang-ci/rust/runs/4309747577?check_suite_focus=true)
<!-- homu: {"type":"BuildFailed","builder_url":"https://github.com/rust-lang-ci/rust/runs/4309747577?check_suite_focus=true","builder_name":"checks-actions"} -->

oli-obk commented:
@bors retry
